src/modules/ekf2/EKF/mag_control.cpp

```diff
@@ -248,63 +252,94 @@ bool Ekf::haglYawResetReq()
 		return false;
 	}
 
+	bool reset_heading = true;
 	bool has_realigned_yaw = false;
 
-	// use yaw estimator if available
-	if (_control_status.flags.gps && isYawEmergencyEstimateAvailable()
-	    && (_mag_counter > 1) // mag LPF available
-	   ) {
+	// reinit mag states
+	const Vector3f mag_I_before_reset = _state.mag_I;
+	const Vector3f mag_B_before_reset = _state.mag_B;
+
+	const float R_MAG = math::max(sq(_params.mag_noise), sq(0.01f));
+
+	// if world magnetic model (inclination, declination, strength) available then use it to reset mag states
+	if (PX4_ISFINITE(_mag_inclination_gps) && PX4_ISFINITE(_mag_declination_gps) && PX4_ISFINITE(_mag_strength_gps)) {
 
-		resetQuatStateYaw(_yawEstimator.getYaw(), _yawEstimator.getYawVar());
+		// use expected earth field to reset states
+		const Vector3f mag_earth_pred = Dcmf(Eulerf(0, -_mag_inclination_gps, _mag_declination_gps))
+						* Vector3f(_mag_strength_gps, 0, 0);
 
-		_information_events.flags.yaw_aligned_to_imu_gps = true;
+		if (isYawEmergencyEstimateAvailable()) {
+			const Dcmf R_to_earth = updateYawInRotMat(_yawEstimator.getYaw(), _R_to_earth);
+			const Dcmf R_to_body = R_to_earth.transpose();
 
-		// if world magnetic model (inclination, declination, strength) available then use it to reset mag states
-		if (PX4_ISFINITE(_mag_inclination_gps) && PX4_ISFINITE(_mag_declination_gps) && PX4_ISFINITE(_mag_strength_gps)) {
-			// use predicted earth field to reset states
-			const Vector3f mag_earth_pred = Dcmf(Eulerf(0, -_mag_inclination_gps, _mag_declination_gps)) * Vector3f(_mag_strength_gps, 0, 0);
-			_state.mag_I = mag_earth_pred;
+			// mag_B: reset using WMM and yaw estimator, initialize variances (so resetMagHeading will use mag bias)
+			_state.mag_B = mag - (R_to_body * mag_earth_pred);
+			P.uncorrelateCovarianceSetVariance<3>(19, R_MAG);
+			_saved_mag_bf_variance = P.slice<3, 3>(19, 19).diag();
 
-			const Dcmf R_to_body = quatToInverseRotMat(_state.quat_nominal);
-			_state.mag_B = _mag_lpf.getState() - (R_to_body * mag_earth_pred);
+			ECL_INFO("using yaw estimator to reset mag bias [%.3f, %.3f, %.3f] -> [%.3f, %.3f, %.3f]",
+				 (double)mag_B_before_reset(0), (double)mag_B_before_reset(1), (double)mag_B_before_reset(2),
+				 (double)_state.mag_B(0), (double)_state.mag_B(1), (double)_state.mag_B(2)
+				);
 
 		} else {
-			// Use the last magnetometer measurements to reset the field states
-			// calculate initial earth magnetic field states
-			_state.mag_I = _R_to_earth * _mag_lpf.getState();
+			// mag_B: reset
 			_state.mag_B.zero();
 		}
 
-		ECL_DEBUG("resetting mag I: [%.3f, %.3f, %.3f], B: [%.3f, %.3f, %.3f]",
-			(double)_state.mag_I(0), (double)_state.mag_I(1), (double)_state.mag_I(2),
-			(double)_state.mag_B(0), (double)_state.mag_B(1), (double)_state.mag_B(2)
-			);
+		if (reset_heading) {
+			has_realigned_yaw = resetMagHeading(mag);
+		}
 
-		resetMagCov();
+		// mag_I: reset to WMM
+		_state.mag_I = mag_earth_pred;
```

Comment:
If you change this line by 
```
_state.mag_I = _R_to_earth * mag;
```
the whole change indicator diff disappears. So I'm pretty sure it's all good.